### PR TITLE
add support for checking if a struct is empty

### PR DESCRIPTION
**Pull request by Stebalien**
(used by the `omitempty` tag)

Rational: Even if it's slow, we only hit this if the user has *explicitly* asked us to omit the empty struct (via `omitempty`). Not doing this is, IMO, a bug.

### CODE REVIEW

**Comment by Stebalien** [comment]
Fixes: https://github.com/ipfs/go-hamt-ipld/pull/19

File: obj/empty.go [hunk 1]
```diff
@@ -21,22 +17,15 @@ func isEmptyValue(v reflect.Value) bool {
...
-// zeroishness checks on a struct are possible in theory but we're passing for now.
-// There's no easy route exposed from the reflect package, due to a variety of reasons:
-//
-//   - The `DeepEqual` method would almost suffice, but we'd need one that takes
-//     `reflect.Value` instead of `interface{}` params, because we already have
-//     the former, and un/re-boxing those into two `interface{}` values is two
-//     heap mallocs and that's a wildly unacceptable performance overhead for this.
-//     `DeepEqual` calls `deepValueEqual`, which does what we want, but...
-//   - We can't easily copy the `deepValueEqual` method.  It imports the unsafe
-//     package.  This is undesirable because it would reduce the portability of refmt.
```
Comment (Stebalien) [inline obj/empty.go]: Unfortunately, these statements aren't *quite* correct for the reason I stated in my comment: we care about *empty*, not *zero*. We can (and often do) have non-zero "empty" values.

File: obj/empty.go [hunk 2]
```diff
@@ -21,22 +17,15 @@ func isEmptyValue(v reflect.Value) bool {
 		return v.Float() == 0
 	case reflect.Interface, reflect.Ptr:
 		return v.IsNil()
```
Comment (Stebalien) [inline obj/empty.go]: Given the fact that we consider a length-0 slice/array to be "empty" (but non-nil), this check here is slightly odd. For example, an empty map stored in an interface will be considered non-empty (right?).

File: obj/empty.go [hunk 3]
```diff
@@ -21,22 +17,15 @@ func isEmptyValue(v reflect.Value) bool {
...
+	case reflect.Struct:
+		// Note: we can't rely on a *literal* "is zero" check because
+		// non-zero pointers may still point to *empty* things.
+		for i := 0; i < v.NumField(); i++ {
+			if !isEmptyValue(v.Field(i)) {
```
Comment (Stebalien) [inline obj/empty.go]: Given the current definition of `isEmptyValue`, this doesn't need a "seen" map for deduplication. We simply don't follow non-nil pointers.

File: obj/empty.go (hunk 2 above)
Comment (warpfork) [reply obj/empty.go]: I'm not sure I entirely parsed that question correctly, but perhaps the method name is odd and it should actually be called something involving the word "zero"...?

[this is fairly literally identical to the checks used in stdlib json](https://github.com/golang/go/blob/541f9c0345d4ec52d9f4be5913bf8097f687e819/src/encoding/json/encode.go#L314-L330), if that's helpful reference.

File: obj/empty.go [hunk 4]
```diff
@@ -21,22 +17,15 @@ func isEmptyValue(v reflect.Value) bool {
...
-//     the former, and un/re-boxing those into two `interface{}` values is two
-//     heap mallocs and that's a wildly unacceptable performance overhead for this.
-//     `DeepEqual` calls `deepValueEqual`, which does what we want, but...
-//   - We can't easily copy the `deepValueEqual` method.  It imports the unsafe
-//     package.  This is undesirable because it would reduce the portability of refmt.
-//
-// It's possible we could produce a struct isZero method which is *simpler* than
-// `deepValueEqual`, because we can actually just halt on any non-zero pointer,
-// and without any need for following pointers we also need no cycle detection.
-// But this is an exercise I'm leaving for later.  PRs welcome if someone wants it.
```
Comment (warpfork) [inline obj/empty.go]: Heh, I guess this is that PR... :smile_cat:

File: obj/empty.go (hunk 2 above)
Comment (Stebalien) [reply obj/empty.go]: Interesting. Basically, my point is that it's *slightly* looser than an "isZero" check. That is, we admit a non-nil zero-length slice while the zero value is technically `nil`.

Note: I think this is as correct as we can get, it just feels like there are some odd edge-cases.

**Comment by warpfork** [comment]
Okay, I gave it another look, and yep, no complaints.  I wonder if we'll regret not having more full obj marshal tests, but the coverage of the emptyness method seems like it should cover the essence of the new feature.  Mergin'.
